README.md
<img src="https://i.imgur.com/4lNLwdV.png" width="550">

# GitHub Watchman
![Python 2.7 and 3 compatible](https://img.shields.io/pypi/pyversions/github-watchman)
![PyPI version](https://img.shields.io/pypi/v/github-watchman.svg)
![License: MIT](https://img.shields.io/pypi/l/github-watchman.svg)

## About GitHub Watchman

GitHub Watchman is an application that uses the GitHub API to audit GitHub for sensitive data and credentials exposed internally.

### Features
It searches GitHub for internally shared projects and looks at:
- Code
- Commits
- Issues
- Repositories

For the following data:
- GCP keys and service account files
- AWS keys
- Azure keys and service account files
- Google API keys
- Slack API tokens & webhooks
- Private keys (SSH, PGP, any other misc private key)
- Exposed tokens (Bearer tokens, access tokens, client_secret etc.)
- S3 config files
- Passwords in plaintext
- and more

#### Time based searching
You can run GitHub Watchman to look for results going back as far as:
- 24 hours
- 7 days
- 30 days
- All time

This means after one deep scan, you can schedule GitHub Watchman to run regularly and only return results from your chosen timeframe.

### Rules
GitHub Watchman uses custom YAML rules to detect matches in GitHub.

They follow this format:

```yaml
---
filename:
enabled: #[true|false]
meta:
  name:
  author:
  date:
  description: #what the search should find#
  severity: #rating out of 100#
scope: #what to search, any combination of the below#
- code
- commits
- issues
- repositories
test_cases:
  match_cases:
  - #test case that should match the regex#
  fail_cases:
  - #test case that should not match the regex#
strings:
- #search query to use in GitHub#
pattern: #Regex pattern to filter out false positives#
```
There are Python tests to ensure rules are formatted properly and that the Regex patterns work in the `tests` dir

More information about rules, and how you can add your own, is in the file `docs/rules.md`.


### Logging

GitHub Watchman gives the following logging options:
- CSV
- Log file
- Stdout
- TCP stream

When using CSV logging, searches for rules are returned in separate CSV files, for all other methods of logging, results are output in JSON format, perfect for ingesting into a SIEM or other log analysis platform.

For file and TCP stream logging, configuration options need to be passed via `.conf` file or environment variable. See the file `docs/logging.md` for instructions on how to set it up.

If no logging option is given, GitHub Watchman defaults to CSV logging.

## Requirements

### GitHub versions
GitHub Watchman uses the v3 API, and works with GitHub Enterprise Server versions that support the v3 API.

GitHub Watchman also works with GitHub.com (Free, Pro and Team) using the API.

### GitHub personal access token
To run GitHub Watchman, you will need a GitHub personal access token.

You can create a personal access token in the GitHub GUI via Settings -> Developer settings -> Personal access tokens

The token needs no specific scopes assigned, as it searches public repositories in the GitHub instance.

**Note**: Personal access tokens act on behalf of the user who creates them, so I would suggest you create a token using a service account, otherwise the app will have access to your private repositories.

### GitHub URL

You also need to provide the URL of your GitHub instance.

#### Providing token & URL
GitHub Watchman will first try to get the the GitHub token and URL from the environment variables `GITHUB_WATCHMAN_TOKEN` and `GITHUB_WATCHMAN_URL`, if this fails they will be taken from .conf file (see below).

### .conf file
Configuration options can be passed in a file named `watchman.conf` which must be stored in your home directory. The file should follow the YAML format, and should look like below:
```yaml
github_watchman:
  token: abc123
  url: https://github.example.com
  logging:
    file_logging:
      path:
    json_tcp:
      host:
      port:
```
GitHub Watchman will look for this file at runtime, and use the configuration options from here. If you are not using the advanced logging features, leave them blank.

If you are having issues with your .conf file, run it through a YAML linter.

An example file is in `docs/example.conf`

**Note** If you use any other Watchman applications and already have a `watchman.conf` file, just append the conf data for GitHub Watchman to the existing file.

## Installation
Install via pip

`pip install github-watchman`

Or via source

## Usage
GitHub Watchman will be installed as a global command, use as follows:
```
usage: github-watchman [-h] --timeframe {d,w,m,a} --output
                   {csv,file,stdout,stream} [--version] [--all] [--code]
                   [--commits] [--issues] [--repositories]

Monitoring GitHub for sensitive data shared publicly

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --all                 Find everything
  --code                Search code
  --commits             Search commits
  --issues              Search issues
  --repositories        Search merge requests

required arguments:
  --timeframe {d,w,m,a}
                        How far back to search: d = 24 hours w = 7 days, m =
                        30 days, a = all time
  --output {csv,file,stdout,stream}
                        Where to send results


  ```

You can run GitHub Watchman to look for everything, and output to default CSV:

`github-watchman --timeframe a --all`

Or arguments can be grouped together to search more granularly. This will look for commits and milestones for the last 30 days, and output the results to a TCP stream:

`github-watchman --timeframe m --commits --milestones --output stream`

## Other Watchman apps
You may be interested in some of the other apps in the Watchman family:
- [Slack Watchman](https://github.com/PaperMtn/slack-watchman)
- [GitLab Watchman](https://github.com/PaperMtn/gitlab-watchman)

## License
The source code for this project is released under the [GNU General Public Licence](https://www.gnu.org/licenses/licenses.html#GPL). This project is not associated with GitHub.


CHANGELOG.md
## 1.0.1 - 2020-11-x
### Fixed
- Retry added for occasional Requests HTTPSConnectionPool error

## 1.0.0 - 2020-10-01
Release

github_watchman/logger.py
import json
import os
import csv
import logging
import socket
import sys
import logging.handlers
from datetime import datetime
from logging import Logger


class CSVLogger(object):
    def __init__(self):
        self.base_out_path = os.getcwd()
        self.headers = {
            'code': [
                'file_name',
                'file_url',
                'sha',
                'repository_id',
                'repository_node_id',
                'repository_name',
                'repository_url',
                'matches',
            ],
            'commits': [
                'commit_url',
                'sha',
                'comments_url',
                'committer_name',
                'committer_id',
                'committer_email',
                'committer_login',
                'commit_date',
                'message',
                'repository_id',
                'repository_node_id',
                'repository_name',
                'repository_url',
                'matches'
            ],
            'issues': [
                'issue_id',
                'issue_title',
                'issue_body',
                'issue_url',
                'sha',
                'user_login',
                'user_id',
                'state',
                'updated_at',
                'repository_url',
                'matches'
            ],
            'repositories': [
                'repository_id',
                'repository_name',
                'repository_description',
                'repository_url',
                'updated_at',
                'owner_login',
                'owner_id',
                'issue_url',
                'matches'
            ]
        }

    def write_csv(self, filename, scope, input_list):
        """Writes input list to .csv. The headers and output path are passed as variables"""

        path = '{}/{}_{}.csv'.format(self.base_out_path, filename, scope)

        with open(path, mode='w+', encoding='utf-8') as csv_file:
            writer = csv.DictWriter(csv_file, fieldnames=self.headers.get(scope))
            writer.writeheader()
            for data in input_list:
                writer.writerow(data)

        csv_file.close()
        print('CSV written: {}'.format(path))


class LoggingBase(Logger):
    def __init__(self, name='GitHub Watchman'):
        super().__init__(name)
        self.notify_format = logging.Formatter(
            '{"localtime": "%(asctime)s", "level": "NOTIFY", "source": "%(name)s", "scope": "%(scope)s",'
            ' "severity": "%(severity)s", "detection_type": "%(type)s", "detection_data": %(message)s}')
        self.info_format = logging.Formatter(
            '{"localtime": "%(asctime)s", "level": "%(levelname)s", "source": "%(name)s", "message":'
            ' "%(message)s"}')
        self.log_path = ''
        self.logger = logging.getLogger(self.name)
        self.logger.setLevel(logging.DEBUG)


class FileLogger(LoggingBase):
    def __init__(self, log_path):
        LoggingBase.__init__(self)
        self.handler = logging.handlers.WatchedFileHandler(os.path.join(log_path, 'github_watchman.log'))
        self.logger.addHandler(self.handler)

    def log_notification(self, log_data, scope, detect_type, severity):
        self.handler.setFormatter(self.notify_format)
        self.logger.warning(json.dumps(log_data), extra={
            'scope': scope,
            'type': detect_type,
            'severity': severity
        })

    def log_info(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.info(log_data)

    def log_critical(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.critical(log_data)


class StdoutLogger(LoggingBase):
    def __init__(self):
        LoggingBase.__init__(self)
        self.handler = logging.StreamHandler(sys.stdout)
        self.logger.addHandler(self.handler)

    def log_notification(self, log_data, scope, detect_type, severity):
        self.handler.setFormatter(self.notify_format)
        self.logger.warning(json.dumps(log_data), extra={
            'scope': scope,
            'type': detect_type,
            'severity': severity
        })

    def log_info(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.info(log_data)

    def log_critical(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.critical(log_data)


class SocketJSONLogger(object):
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.sock.connect((self.host, self.port))
        except socket.error as error:
            print(error)

    def send(self, data):
        try:
            self.sock.sendall(bytes(data, encoding="utf-8"))
        except Exception as e:
            print(e)

    def log_notification(self, log_data, scope, detect_type, severity):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'NOTIFY',
            'source': 'GitHub Watchman',
            'scope': scope,
            'severity': severity,
            'detection_type': detect_type,
            'detection_data': log_data
        }) + '\n'
        self.send(message)

    def log_info(self, log_data):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'INFO',
            'source': 'GitHub Watchman',
            'message': log_data
        }) + '\n'
        self.send(message)

    def log_critical(self, log_data):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'CRITICAL',
            'source': 'GitHub Watchman',
            'message': log_data
        }) + '\n'
        self.send(message)


github_watchman/__about__.py
# GitHub Watchman
# Copyright (C) 2020  PaperMtn
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

__all__ = [
    '__title__',
    '__summary__',
    '__uri__',
    '__version__',
    '__author__',
    '__email__',
    '__license__',
]

__title__ = 'GitHub Watchman'
__summary__ = 'Monitoring GitHub for sensitive data shared publicly'
__uri__ = 'https://github.com/PaperMtn/github-watchman'
__version__ = '1.0.1'
__author__ = 'PaperMtn'
__email__ = 'papermtn@protonmail.com'
__license__ = 'GPL-3.0'
__copyright__ = '2020 {}'.format(__author__)


github_watchman/__init__.py
import builtins
import argparse
import os
import yaml
import time
from pathlib import Path
from datetime import date
from colorama import init, deinit
from termcolor import colored

import github_watchman.github_wrapper as github
import github_watchman.__about__ as a
import github_watchman.config as cfg
import github_watchman.logger as logger


RULES_PATH = (Path(__file__).parent / 'rules').resolve()
OUTPUT_LOGGER = ''


def validate_conf(path):
    """Check the file watchman.conf exists"""

    if os.environ.get('GITHUB_WATCHMAN_TOKEN') and os.environ.get('GITHUB_WATCHMAN_URL'):
        return True
    if os.path.exists(path):
        with open(path) as yaml_file:
            return yaml.safe_load(yaml_file).get('github_watchman')


def search(github_connection, rule, tf, scope):
    if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
        print = OUTPUT_LOGGER.log_info
    else:
        print = builtins.print
    try:
        if scope == 'code':
            print(colored('Searching for {} in {}'.format(rule.get('meta').get('name'),
                                                          'code'), 'yellow'))

            code = github.search_code(github_connection, OUTPUT_LOGGER, rule, tf)
            if code:
                if isinstance(OUTPUT_LOGGER, logger.CSVLogger):
                    OUTPUT_LOGGER.write_csv('exposed_{}'.format(rule.get('filename').split('.')[0]),
                                            'code',
                                            code)
                else:
                    for log_data in code:
                        OUTPUT_LOGGER.log_notification(log_data, 'code', rule.get('meta').get('name'),
                                                       rule.get('meta').get('severity'))
                    print('Results output to log')

        if scope == 'commits':
            print(colored('Searching for {} in {}'.format(rule.get('meta').get('name'),
                                                          'commits'), 'yellow'))

            commits = github.search_commits(github_connection, OUTPUT_LOGGER, rule, tf)
            if commits:
                if isinstance(OUTPUT_LOGGER, logger.CSVLogger):
                    OUTPUT_LOGGER.write_csv('exposed_{}'.format(rule.get('filename').split('.')[0]),
                                            'commits',
                                            commits)
                else:
                    for log_data in commits:
                        OUTPUT_LOGGER.log_notification(log_data, 'commits', rule.get('meta').get('name'),
                                                       rule.get('meta').get('severity'))
                    print('Results output to log')

        if scope == 'issues':
            print(colored('Searching for {} in {}'.format(rule.get('meta').get('name'),
                                                          'issues'), 'yellow'))

            issues = github.search_issues(github_connection, OUTPUT_LOGGER, rule, tf)
            if issues:
                if isinstance(OUTPUT_LOGGER, logger.CSVLogger):
                    OUTPUT_LOGGER.write_csv('exposed_{}'.format(rule.get('filename').split('.')[0]),
                                            'issues',
                                            issues)
                else:
                    for log_data in issues:
                        OUTPUT_LOGGER.log_notification(log_data, 'issues', rule.get('meta').get('name'),
                                                       rule.get('meta').get('severity'))
                    print('Results output to log')

        if scope == 'repositories':
            print(colored('Searching for {} in {}'.format(rule.get('meta').get('name'),
                                                          'repositories'), 'yellow'))

            repositories = github.search_repositories(github_connection, OUTPUT_LOGGER, rule, tf)
            if repositories:
                if isinstance(OUTPUT_LOGGER, logger.CSVLogger):
                    OUTPUT_LOGGER.write_csv('exposed_{}'.format(rule.get('filename').split('.')[0]),
                                            'repositories',
                                            repositories)
                else:
                    for log_data in repositories:
                        OUTPUT_LOGGER.log_notification(log_data, 'wiki_blobs', rule.get('meta').get('name'),
                                                       rule.get('meta').get('severity'))
                    print('Results output to log')
    except Exception as e:
        if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = OUTPUT_LOGGER.log_critical
        else:
            print = builtins.print

        print(colored(e, 'red'))


def load_rules():
    rules = []
    try:
        for file in os.scandir(RULES_PATH):
            if file.name.endswith('.yaml'):
                with open(file) as yaml_file:
                    rule = yaml.safe_load(yaml_file)
                    if rule.get('enabled'):
                        rules.append(rule)
        return rules
    except Exception as e:
        if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = OUTPUT_LOGGER.log_critical
        else:
            print = builtins.print

        print(colored(e, 'red'))


def main():
    global OUTPUT_LOGGER
    try:
        init()

        parser = argparse.ArgumentParser(description=a.__summary__)
        required = parser.add_argument_group('required arguments')
        required.add_argument('--timeframe', choices=['d', 'w', 'm', 'a'], dest='time',
                              help='How far back to search: d = 24 hours w = 7 days, m = 30 days, a = all time',
                              required=True)
        required.add_argument('--output', choices=['csv', 'file', 'stdout', 'stream'], dest='logging_type',
                              help='Where to send results', required=True)
        parser.add_argument('--version', action='version',
                            version='github-watchman {}'.format(a.__version__))
        parser.add_argument('--all', dest='everything', action='store_true',
                            help='Find everything')
        parser.add_argument('--code', dest='code', action='store_true',
                            help='Search code')
        parser.add_argument('--commits', dest='commits', action='store_true',
                            help='Search commits')
        parser.add_argument('--issues', dest='issues', action='store_true',
                            help='Search issues')
        parser.add_argument('--repositories', dest='repositories', action='store_true',
                            help='Search merge requests')

        args = parser.parse_args()
        tm = args.time
        everything = args.everything
        code = args.code
        commits = args.commits
        repositories = args.repositories
        issues = args.issues
        logging_type = args.logging_type

        if tm == 'd':
            tf = cfg.DAY_TIMEFRAME
        elif tm == 'w':
            tf = cfg.WEEK_TIMEFRAME
        elif tm == 'm':
            tf = cfg.MONTH_TIMEFRAME
        else:
            tf = cfg.ALL_TIME
        conf_path = '{}/watchman.conf'.format(os.path.expanduser('~'))

        if not validate_conf(conf_path):
            raise Exception(
                colored('GITHUB_WATCHMAN_TOKEN environment variable or watchman.conf file not detected. '
                        '\nEnsure environment variable is set or a valid file is located in your home '
                        'directory: {} ', 'red')
                .format(os.path.expanduser('~')))
        else:
            config = validate_conf(conf_path)
            connection = github.initiate_github_connection()

        if logging_type:
            if logging_type == 'file':
                if os.environ.get('GITHUB_WATCHMAN_LOG_PATH'):
                    OUTPUT_LOGGER = logger.FileLogger(os.environ.get('GITHUB_WATCHMAN_LOG_PATH'))
                elif config.get('logging').get('file_logging').get('path') and \
                        os.path.exists(config.get('logging').get('file_logging').get('path')):
                    OUTPUT_LOGGER = logger.FileLogger(log_path=config.get('logging').get('file_logging').get('path'))
                else:
                    print('No config given, outputting github_watchman.log file to home path')
                    OUTPUT_LOGGER = logger.FileLogger(log_path=os.path.expanduser('~'))
            elif logging_type == 'stdout':
                OUTPUT_LOGGER = logger.StdoutLogger()
            elif logging_type == 'stream':
                if os.environ.get('GITHUB_WATCHMAN_HOST') and os.environ.get('GITHUB_WATCHMAN_PORT'):
                    OUTPUT_LOGGER = logger.SocketJSONLogger(os.environ.get('GITHUB_WATCHMAN_HOST'),
                                                            os.environ.get('GITHUB_WATCHMAN_PORT'))
                elif config.get('logging').get('json_tcp').get('host') and \
                        config.get('logging').get('json_tcp').get('port'):
                    OUTPUT_LOGGER = logger.SocketJSONLogger(config.get('logging').get('json_tcp').get('host'),
                                                            config.get('logging').get('json_tcp').get('port'))
                else:
                    raise Exception("JSON TCP stream selected with no config")
            else:
                OUTPUT_LOGGER = logger.CSVLogger()
        else:
            print('No logging option selected, defaulting to CSV')
            OUTPUT_LOGGER = logger.CSVLogger()

        now = int(time.time())
        today = date.today().strftime('%Y-%m-%d')
        start_date = time.strftime('%Y-%m-%d', time.localtime(now - tf))

        if not isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = builtins.print
            print(colored('''
              #####          #     #                                 
             #     # # ##### #     # #    # #####                    
             #       #   #   #     # #    # #    #                   
             #  #### #   #   ####### #    # #####                    
             #     # #   #   #     # #    # #    #                   
             #     # #   #   #     # #    # #    #                   
              #####  #   #   #     #  ####  #####                    
                                                                     
             #     #                                                 
             #  #  #   ##   #####  ####  #    # #    #   ##   #    # 
             #  #  #  #  #    #   #    # #    # ##  ##  #  #  ##   # 
             #  #  # #    #   #   #      ###### # ## # #    # # #  # 
             #  #  # ######   #   #      #    # #    # ###### #  # # 
             #  #  # #    #   #   #    # #    # #    # #    # #   ## 
              ## ##  #    #   #    ####  #    # #    # #    # #    #
                                                                 ''', 'magenta'))
            print('Version: {}\n'.format(a.__version__))
            print('Searching from {} to {}'.format(start_date, today))
            print('Importing rules...')
            rules_list = load_rules()
            print('{} rules loaded'.format(len(rules_list)))
        else:
            OUTPUT_LOGGER.log_info('GitHub Watchman started execution')
            OUTPUT_LOGGER.log_info('Version: {}'.format(a.__version__))
            OUTPUT_LOGGER.log_info('Importing rules...')
            rules_list = load_rules()
            OUTPUT_LOGGER.log_info('{} rules loaded'.format(len(rules_list)))
            print = OUTPUT_LOGGER.log_info

        if everything:
            print(colored('Getting everything...', 'magenta'))
            for rule in rules_list:
                if 'code' in rule.get('scope'):
                    search(connection, rule, tf, 'code')
                if 'commits' in rule.get('scope'):
                    search(connection, rule, tf, 'commits')
                if 'issues' in rule.get('scope'):
                    search(connection, rule, tf, 'issues')
                if 'repositories' in rule.get('scope'):
                    search(connection, rule, tf, 'repositories')
        else:
            if code:
                print(colored('Searching blobs', 'magenta'))
                for rule in rules_list:
                    if 'code' in rule.get('scope'):
                        search(connection, rule, tf, 'code')
            if commits:
                print(colored('Searching commits', 'magenta'))
                for rule in rules_list:
                    if 'commits' in rule.get('scope'):
                        search(connection, rule, tf, 'commits')
            if issues:
                print(colored('Searching issues', 'magenta'))
                for rule in rules_list:
                    if 'issues' in rule.get('scope'):
                        search(connection, rule, tf, 'issues')
            if repositories:
                print(colored('Searching repositories', 'magenta'))
                for rule in rules_list:
                    if 'repositories' in rule.get('scope'):
                        search(connection, rule, tf, 'repositories')

        print(colored('++++++Audit completed++++++', 'green'))

        deinit()

    except Exception as e:
        if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = OUTPUT_LOGGER.log_critical
        else:
            print = builtins.print

        print(colored(e, 'red'))


if __name__ == '__main__':
    main()
